Send Gas Tooltip

danjm opened the pull request:
4 - Send Ether - Add clickable "customize" toggle option to Gas Fee Field
5 - Send Ether - Gas Tooltip UI
6 - Send Ether - Gas Tooltip State

_Update - screenshots added (8-29)_

Full screen
<img width="398" alt="screen shot 2017-08-29 at 1 40 37 pm" src="https://user-images.githubusercontent.com/7499938/29831464-d7642302-8cbf-11e7-959b-1855cdd54a41.png">
<img width="388" alt="screen shot 2017-08-29 at 1 40 44 pm" src="https://user-images.githubusercontent.com/7499938/29831465-d76a2216-8cbf-11e7-92df-ef2ca0ddc1b0.png">

Extension
<img width="335" alt="screen shot 2017-08-29 at 1 41 14 pm" src="https://user-images.githubusercontent.com/7499938/29831466-d76bb22a-8cbf-11e7-85d2-5d0c71467c5d.png">

Popup
<img width="319" alt="screen shot 2017-08-29 at 1 42 44 pm" src="https://user-images.githubusercontent.com/7499938/29831523-f90e4a5a-8cbf-11e7-82ab-97e1a884b950.png">

danjm committed:
Add 'customize' text to gas input.

danjm committed:
Toggling tooltip.

danjm committed:
Refactor tooltip to remove external lib; tooltip now updating gas fee in parent.

danjm committed:
Tooltip closing on click outside.

danjm committed:
Updating current currency from send form.

danjm committed:
Fully connect gas data in send form and tooltip to state; final styling improvements.
Fully connect gas fields in send form and tooltip to state

danjm committed:
Final styling improvements for tooltip; plus a few for send form.

sdtsui commented on ui/app/components/gas-tooltip.js:
```diff
@@ -0,0 +1,136 @@
+  const { position, title, children, className, isOpen } = this.props
+  const { gasPrice, gasLimit } = this.state
+
+  this.manageListeners()
```
Is there a reason why we need to manage listeners on each render?
-  "On each render" seems to break isolation of concerns. A lifecycle method (like `componentWillReceiveProps`, or a mount/unmount lifecylce method perhaps?
- I see that we use it to change `onClickOutside`, can you elaborate as to why this needs to be in `render`?

sdtsui commented on ui/app/components/gas-tooltip.js:
```diff
@@ -0,0 +1,136 @@
+    : null
+}
+
+GasTooltip.prototype.manageListeners = function () {
```
see above - might not need this function, just the logic inside a lifecycle method

sdtsui commented on ui/app/send.js:
```diff
@@ -53,9 +55,11 @@ function SendTransactionScreen () {
       amount: '0.0001', // see L544
       gasPrice: '4a817c800',
       gas: '0x7b0d',
+      gasFee: ((parseInt('0x7b0d', 16) * parseInt('4a817c800', 16)) / 1000000000).toFixed(10),
```
Would readability be improved by moving this out into a static method, or util function?

sdtsui reviewed (commented):
@danjm - thanks.

- This is a pretty good break point for cleanup, as it's a new component. Can we get some linting added (use `stylelint` for scss, and `npm run lint` for JS - they're all in `package.json`)? Feel free to do any extra component cleanup as well (renaming funcs, new components, code reorganization)
- Can we add some screenshots to this PR?

chikeichan commented on package.json:
```diff
@@ -121,6 +121,7 @@
     "react-redux": "^5.0.5",
     "react-select": "^1.0.0-rc.2",
     "react-simple-file-input": "^1.0.0",
+    "react-tooltip": "^3.3.0",
```
Are we still using this?

chikeichan commented on ui/app/components/gas-tooltip.js:
```diff
@@ -0,0 +1,136 @@
+  }
+}
+
+function isDescendant (parent, child) {
```
Let's sycn up tomorrow. I see that you are trying to duplicate `react-tooltip` functionality directly on DOM nodes. This is generally undesirable in react. We can jam on better approach tomorrow.

I will send you a link on how we handled popover the react way in our other project.

chikeichan commented on ui/app/css/itcss/components/send.scss:
```diff
@@ -70,4 +79,117 @@
 
-.send-screen-bolt-icon {}
\ No newline at end of file
+.send-screen-bolt-icon {}
```
remove

danjm replied:
removed in https://github.com/danjm/metamask-extension/tree/NewUI-flat-lint-fixes, which I will make a new PR for after this one

chikeichan commented on ui/app/css/itcss/components/send.scss:
```diff
@@ -70,4 +79,117 @@
+  padding: 13px 19px;
+  font-size: 16px;
+  border-radius: 4px;
+  font-family: 'Avenir Next';
```
is this font available in metamask?

danjm replied:
I've replaced it with a new font that I added "Lato" which is a near equivalent alternative.

danjm committed:
Styles and behaviour correct

danjm committed:
Refactor for clean handling of tooltip close.

danjm committed:
Remove unused dependency.

danjm committed:
Bind SendTransactionScreen methods to this in constructor.

danjm committed:
Using eth balance component to ensure proper rounding.

danjm committed:
Move getTaxBN and bnMultiplyByFraction to util.js

danjm committed:
Use hex values only in send.js to handle limit and price; GasTooltip accepts and returns values as hex (allows user to enter floats)

danjm committed:
Center tooltip on small screen size by using flexbox

danjm commented:
> This is a pretty good break point for cleanup, as it's a new component. Can we get some linting added (use stylelint for scss, and npm run lint for JS - they're all in package.json)? Feel free to do any extra component cleanup as well (renaming funcs, new components, code reorganization)
> Can we add some screenshots to this PR?

I have moved most of the cleanup to a new branch so that it can be reviewed separately: https://github.com/danjm/metamask-extension/tree/send-gas-tooltip-cleanup I'll make a new PR  for that branch once this PR is merged

I have moved most of the linting fixes to a new branch so that they can be reviewed separately: https://github.com/danjm/metamask-extension/tree/NewUI-flat-lint-fixes. In this case there were a lot of lint errors not related to this PR, but fixes the lint errors in the PR was greatly simplified by fixing all the lint errors. I'll make a new PR  for that branch once this PR is merged

danjm committed:
Replace Avenir Next with alternative.

chikeichan commented on ui/app/send.js:
```diff
@@ -51,12 +58,23 @@ function SendTransactionScreen () {
+  this.onSubmit = this.onSubmit.bind(this)
+  this.recipientDidChange = this.recipientDidChange.bind(this)
+  this.setCurrentCurrency = this.setCurrentCurrency.bind(this)
+  this.setCurrentCurrency = this.setCurrentCurrency.bind(this)
```
Some of these are binded twice

chikeichan commented on ui/app/send.js:
```diff
@@ -195,6 +225,23 @@ SendTransactionScreen.prototype.render = function () {
+            onFeeChange: ({gasLimit, gasPrice}) => {
+              this.setState({
+                newTx: Object.assign(
+                  this.state.newTx,
```
I believe this would mutate state.newTx. we should use es6 spread to ensure immutability

chikeichan commented on ui/app/send.js:
```diff
@@ -212,9 +259,39 @@ SendTransactionScreen.prototype.render = function () {
+                }
+            }),
+            h('div.send-screen-gas-input-customize', {
+              onClick: () => this.toggleTooltip.bind(this)(!this.state.tooltipIsOpen),
```
No need to bind when you are using arrow function

chikeichan commented on ui/app/components/gas-tooltip.js:
```diff
@@ -0,0 +1,107 @@
+
+  return h('div', {
+    style: {
+      display: 'flex',
```
what's the rationale behind inline vs sass? Would love to minimize inline style as much as possible so we have consistent css convention. Properties like flex and transform are better in sass as well because of autoprefixing

danjm committed:
Minor clean up.
